File: android/EyeTimeTrackerAndroid/src/com/eyetimetracker/android/ContinuousReminderBaseline.java
package com.eyetimetracker.android;

public final class ContinuousReminderBaseline {
    private ContinuousReminderBaseline() {
    }

    public static long resolve(
            ReminderRuntimeState localState,
            ReminderRuntimeState peerState,
            boolean peerOnline) {
        if (localState == null
                || !localState.isCounting
                || localState.currentSessionStartedUnixSeconds <= 0L) {
            return 0L;
        }

        if (peerOnline
                && peerState != null
                && peerState.isCounting
                && peerState.currentSessionStartedUnixSeconds > 0L) {
            return Math.min(
                    localState.currentSessionStartedUnixSeconds,
                    peerState.currentSessionStartedUnixSeconds);
        }

        return localState.currentSessionStartedUnixSeconds;
    }
}

File: src/EyeTimeTracker.Core/Reminders/ContinuousReminderBaseline.cs
using EyeTimeTracker.Core.Sync;

namespace EyeTimeTracker.Core.Reminders;

public static class ContinuousReminderBaseline
{
    private const long MaxPlausibleSessionSeconds = 24 * 60 * 60; // 24 小时

    public static long Resolve(
        ReminderRuntimeState localState,
        ReminderRuntimeState? peerState,
        bool peerOnline,
        long nowUnixSeconds = 0)
    {
        if (localState is null
            || !localState.IsCounting
            || localState.CurrentSessionStartedUnixSeconds <= 0)
        {
            return 0;
        }

        var now = nowUnixSeconds > 0
            ? nowUnixSeconds
            : DateTimeOffset.Now.ToUnixTimeSeconds();

        var localStart = localState.CurrentSessionStartedUnixSeconds;
        if (now - localStart > MaxPlausibleSessionSeconds)
        {
            return 0;
        }

        if (peerOnline
            && peerState is not null
            && peerState.IsCounting
            && peerState.CurrentSessionStartedUnixSeconds > 0)
        {
            var peerStart = peerState.CurrentSessionStartedUnixSeconds;
            if (now - peerStart > MaxPlausibleSessionSeconds)
            {
                return localStart;
            }

            return Math.Min(localStart, peerStart);
        }

        return localStart;
    }
}
